tools/calyx-pass/src/pass_explorer.rs

```
@@ -0,0 +1,302 @@
+                new_self.accept()?;
+                } else {
+                    new_self.skip()?;
+                }
+            }
+        }
+
+        Ok(new_self)
+    }
+
+    /// The pass most recently applied, if one exists.
+    pub fn last_pass(&self) -> Option<String> {
+        self.passes_applied
+            .last()
+            .map(|index| self.passes[*index].clone())
+    }
+
+    /// The pass staged to be applied, if one exists.
+    pub fn incoming_pass(&self) -> Option<String> {
+        self.passes.get(self.index as usize).cloned()
+    }
+
+    /// An association of each pass with its current exploration status.
+    pub fn current_pass_application(
+        &self,
+    ) -> Vec<(String, PassApplicationStatus)> {
+        let mut result = vec![];
+        let mut walk_applied = 0;
+        for i in 0..self.passes.len() {
+            let pass = self.passes[i].clone();
+            if walk_applied < self.passes_applied.len()
+                && i == self.passes_applied[walk_applied]
+            {
+                result.push((pass, PassApplicationStatus::Applied));
+                walk_applied += 1;
+            } else if i < (self.index as usize) {
+                result.push((pass, PassApplicationStatus::Skipped));
+            } else if i == (self.index as usize) {
+                result.push((pass, PassApplicationStatus::Incoming));
+            } else {
+                result.push((pass, PassApplicationStatus::Future));
+            }
+        }
+        result
+    }
+
+    /// Produces a printable diff showing how the
+    /// [`PassExplorer::incoming_pass`] will transform the current file state.
+    pub fn review(
+        &mut self,
+        component: Option<String>,
+    ) -> std::io::Result<Option<String>> {
+        self.ensure_inc_file_exists()?;
+        let mut last_file_content = fs::read_to_string(self.last_file())
+            .expect("Could not read the last file");
```

`expect` will crash the tool instead of gracefully showing an error. Why not bubble up the error message?